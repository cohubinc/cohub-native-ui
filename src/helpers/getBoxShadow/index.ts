import { ElevationLevel } from "src/definitions/enums/BoxShadow";
import { ViewStyle } from "react-native";

type ShadowStyles = Required<
  Pick<
    ViewStyle,
    | "shadowColor"
    | "shadowOffset"
    | "shadowOpacity"
    | "shadowRadius"
    | "elevation"
  >
>;

export default function getBoxShadow(elevation: ElevationLevel) {
  if (elevation === 0) return null;

  return shadowMap[elevation];
}

type TShadowMap = { [elevation in ElevationLevel]: ShadowStyles };
const shadowMap: Omit<TShadowMap, 0> = {
  1: {
    shadowColor: "rgba(0, 0, 0, 0.14)",
    shadowOffset: { height: 1, width: 0 },
    shadowOpacity: 1,
    shadowRadius: 1,
    elevation: 1
  },
  2: {
    shadowColor: "rgba(0, 0, 0, 0.15)",
    shadowOffset: { height: 1, width: 0 },
    shadowOpacity: 1,
    shadowRadius: 2,
    elevation: 2
  },
  3: {
    shadowColor: "rgba(0, 0, 0, 0.12)",
    shadowOffset: { height: 1, width: 0 },
    shadowOpacity: 1,
    shadowRadius: 8,
    elevation: 3
  },
  8: {
    shadowColor: "rgba(0, 0, 0, 0.1)",
    shadowOffset: { height: 5, width: 0 },
    shadowOpacity: 1,
    shadowRadius: 5,
    elevation: 8
  },
  16: {
    shadowColor: "rgba(0, 0, 0, 0.2)",
    shadowOffset: { height: 8, width: 0 },
    shadowOpacity: 1,
    shadowRadius: 10,
    elevation: 16
  },
  24: {
    shadowColor: "rgba(0, 0, 0, 0.2)",
    shadowOffset: { height: 11, width: 0 },
    shadowOpacity: 1,
    shadowRadius: 15,
    elevation: 24
  }
};
