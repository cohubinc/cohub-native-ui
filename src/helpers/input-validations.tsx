import isNumber from "lodash/isNumber";

type TValidator<T> = (value?: T) => string | undefined;

export function composeValidators<T = any>(
  ...validators: Array<TValidator<T>>
) {
  return (value?: T) => {
    validators.reduce<string | undefined>(
      (error, validator) => error || validator(value),
      undefined
    );
  };
}

export const required = (value?: string | null) =>
  value && value.trim() ? undefined : "Required";

export const requiredNumber = (value?: number | null) =>
  value && isNumber(value) ? undefined : "Required";

export const minLength = (min: number) => (
  value: string | number | object = ""
) =>
  typeof value === "string" && value.length < min
    ? `Should be at least ${min} characters long`
    : undefined;

export const minValue = (min: number) => (value?: string | number | null) =>
  typeof value !== "string" && value && (isNaN(value) || value >= min)
    ? undefined
    : `Should be greater than ${min}`;

export const length = (valLength: number) => (value = "") => {
  return value.length === valLength
    ? undefined
    : `Should be ${valLength} characters long`;
};

export const email = composeValidators(
  (value = "") =>
    typeof value === "string" && charsArePresent(value, "@", ".")
      ? undefined
      : "Should be a valid email",
  minLength(4)
);

export const lengthRange = (min: number, max: number) => (v = "") => {
  if (v.length >= min && v.length <= max) {
    return undefined;
  }

  return `Should be between ${min} and ${max} characters long`;
};

export const isInt = (value?: string | null) => {
  if (!value) {
    return false;
  }

  const isNum = /^\d+$/.test(value);
  const parsedVal = isNum && Number.parseFloat(value);

  return parsedVal && Number.isInteger(parsedVal);
};

const charsArePresent = (string: string, ...chars: string[]) =>
  chars.every(char => string.includes(char));
