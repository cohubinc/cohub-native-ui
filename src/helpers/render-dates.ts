import moment from "moment";

type TFormat =
  | "monthDayYear"
  | "monthDayYearShort"
  | "dateWithTime"
  | "dateWithTimeShort"
  | "monthShort"
  | "dateShort"
  | "tabularDate"
  | "input";

// Use like -> renderDate("dateWithTimeShort")(item.created_at)
export const renderDate = (format: TFormat) => (dateTime?: string) => {
  if (!dateTime) return "";

  return moment(dateTime).format(formatMap[format]);
};

type TFormatMap = { [Key in TFormat]: string };
const formatMap: TFormatMap = {
  monthDayYear: "MMM Do[,] YYYY",
  monthDayYearShort: "MMM D[,] YYYY",
  dateWithTime: "MMM Do[,] YYYY [at] h:mma",
  dateWithTimeShort: "MM/DD/YYYY [at] h:mma",
  monthShort: "MMM",
  dateShort: "MM/DD/YY",
  tabularDate: "MM/DD/YY",
  input: "MM-DD-YYYY"
};
