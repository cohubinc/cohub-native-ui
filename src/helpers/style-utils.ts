export const paddingVertical = (top: string | number, bottom = top) => ({
  paddingTop: top,
  paddingBottom: bottom
});

export const paddingHorizontal = (left: string | number, right = left) => ({
  paddingLeft: left,
  paddingRight: right
});

export const marginVertical = (top: string | number, bottom = top) => ({
  marginTop: top,
  marginBottom: bottom
});

export const marginHorizontal = (left: string | number, right = left) => ({
  marginLeft: left,
  marginRight: right
});

export const size = (height: string | number, width = height) => ({
  height: width,
  width: height
});
