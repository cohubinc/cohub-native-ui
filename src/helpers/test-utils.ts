import { StyleProp, ViewStyle } from "react-native";

export function isLineLoaderVisible(loader: any) {
  return !(loader?.props?.style?.pop().opacity === 0);
}

export function findActiveStyle(
  style: StyleProp<ViewStyle>,
  property: string
): string | undefined {
  if (!style) {
    return;
  }

  if (Array.isArray(style)) {
    return findActiveStyle(
      style.reverse().find((s) => !!findActiveStyle(s, property)),
      property
    );
  }

  if (typeof style === "object") {
    return (style as any)[property as any];
  }
}
