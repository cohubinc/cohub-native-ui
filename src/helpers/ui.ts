export const truncateString = (length: number, separator: string = "...") => (
  string: string
) =>
  length < string.length
    ? (string.slice(0, length) as any).trimEnd() + separator
    : string;

export function childIsVisible(parent: Element, child: Element, offset = 120) {
  if (!parent || !child) return true;

  const parentRect = parent.getBoundingClientRect();

  const parentViewableArea = {
    height: parent.clientHeight,
    width: parent.clientWidth
  };

  const childRect = child.getBoundingClientRect();

  // Is the child viewable?
  const childViewable =
    childRect.top >= parentRect.top &&
    childRect.top <= parentRect.top + parentViewableArea.height - offset;

  return childViewable;
}

export function stringifiedObjectValues(obj: string, separator: string = ", ") {
  return Object.values(JSON.parse(obj)).join(separator);
}
