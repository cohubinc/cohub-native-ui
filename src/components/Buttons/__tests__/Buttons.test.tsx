import React from "react";
import { Text } from "react-native";
import { render, fireEvent } from "@testing-library/react-native";

import Button from "../";
import { isLineLoaderVisible, findActiveStyle } from "src/helpers/test-utils";

const onPress = () => null;

describe("Buttons", () => {
  describe("base Button", () => {
    it("Passing label text as a string renders", () => {
      const { getByText } = render(
        <Button onPress={onPress} label="Base Button" />
      );

      expect(getByText("Base Button")).toBeTruthy();
    });

    it("passing label text as a ReactNode renders", () => {
      const { getByText } = render(
        <Button onPress={onPress} label={<Text>Node Text</Text>} />
      );

      expect(getByText("Node Text")).toBeTruthy();
    });

    it("loading prop true works as expected", () => {
      const { getByTestId } = render(
        <Button onPress={onPress} loading label="Base Button" />
      );

      const loader = getByTestId("line-loader");

      expect(isLineLoaderVisible(loader)).toBeTruthy();
    });

    it("loading prop false works as expected", () => {
      const { getByTestId } = render(
        <Button onPress={onPress} loading={false} label="Base Button" />
      );

      const loader = getByTestId("line-loader");

      expect(isLineLoaderVisible(loader)).toBe(false);
    });

    it("elevationLevel prop 3 works as expected", () => {
      const { getByTestId } = render(
        <Button
          onPress={onPress}
          disabled
          elevationLevel={3}
          label="Base Button"
        />
      );

      const style = getByTestId("button-container")?.props?.style;

      const hasShadow = findActiveStyle(style, "shadowColor");

      expect(hasShadow).toBeTruthy();
    });

    it("onPress prop works as expected", () => {
      const onPressSpy = jest.fn();

      const { getByLabelText } = render(
        <Button onPress={onPressSpy} label="Base Button" />
      );

      fireEvent.press(getByLabelText("Base Button"));

      expect(onPressSpy.mock.calls.length).toBe(1);
    });

    it("disabled prop true works as expected", () => {
      const onPressSpy = jest.fn();

      const { getByLabelText } = render(
        <Button onPress={onPressSpy} disabled label="Base Button" />
      );

      fireEvent.press(getByLabelText("Base Button"));

      expect(onPressSpy.mock.calls.length).toBe(0);
    });
  });
});
