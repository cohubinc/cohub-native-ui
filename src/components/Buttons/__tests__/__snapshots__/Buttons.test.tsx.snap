// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buttons renders 1`] = `
<Buttons>
  <Base>
    <button
      className="CohubButton "
      style={
        Object {
          "backgroundColor": "hsl(0, 0%, 77%)",
          "boxShadow": "0 2px 3px hsla(0, 0%, 0%, 0.2), inset 0px 2px 0px hsl(0, 0%, 92%)",
        }
      }
    >
      <div
        className="button-text relative"
      >
        <Typography
          color="#EFF7EE"
          secondary={true}
          style={
            Object {
              "opacity": 1,
              "transition": "opacity 150ms ease-in",
            }
          }
          uppercase={true}
        >
          <Component
            color="#EFF7EE"
            secondary={true}
            style={
              Object {
                "opacity": 1,
                "transition": "opacity 150ms ease-in",
              }
            }
            uppercase={true}
          >
            <span
              className=""
              data-qa="text"
              style={
                Object {
                  "color": "#EFF7EE",
                  "fontFamily": "Akkurat-Mono",
                  "fontSize": "14px",
                  "fontWeight": 400,
                  "opacity": 1,
                  "textAlign": "initial",
                  "textTransform": "uppercase",
                  "transition": "opacity 150ms ease-in",
                }
              }
            >
              <span>
                push me
              </span>
            </span>
          </Component>
        </Typography>
      </div>
    </button>
  </Base>
</Buttons>
`;
